Dynatrace/dynatrace-operator
# Rework e2e tests

## Description
The structure of our e2e tests have been redone, to make them less flaky. (also many fixes were introduced)

Now instead of redeploying the operator for each "feature", instead we group the features into "scenarios".
Each scenarios is a type of install of the operator.
- `standard`: installing the latest operator, according to the branch we are running the tests
- `istio`:  installing `standard` + istio is injected into the dynatrace namespace, features that require network restrictions can be found here.
- `release`:   installing a previously released operator version, this is only used for the "upgrade" test, and probably it will stay that way.

By doing this, we are also speeding up the test by a lot, also switching between modes/custom-resources should be something that the operator can handle.

Changes to the make targets:
- now we are using labels to run a single feature test, these labels need to be defined when implementing the feature test
- no more `run_all.sh` 🎉
- limitation of the `/debug` targets:
     - it will **never** clean up the operator, even if everything passed, keep this in mind. (there was no reasonable way to get over this)
     - this is why there is no `test/e2e/debug`, because switching between the scenarios can/will cause problems duo to lack of cleanup

Some refactors:
- Removed all the `Builder` patterns, and replaced them with the "functional options pattern"
   - Why?: Its a more go-like pattern, and the `Builder` pattern enabled usage patterns that were convoluted, messy and prone to error. (example: reusing a builder)
- Some functions were changed to be `env.Func` from `feature.Func`, so they can be used during the `Setup` and `Finish`
    - Its easy to make an `feature.Func` from an `env.Func` on the fly, wrote a simple helper for that. (the other way is not that easy) 

## How can this be tested?

`make test/e2e/....` every command
MAKE SURE YOUR CLUSTER IS CLEAN, otherwise it might get flaky

## Checklist
